File: 프로그래머스/2/12973. 짝지어 제거하기/짝지어 제거하기.java
import java.util.*;

class Solution
{
    public int solution(String s)
    {
        /*
        문자열 길이 1 -> return 0
        
        스택에 맨앞 문자 넣고
        맨앞 제외 문자열 순회 (c)
            스택빔 -> 그냥 넣음
            else
                스택 peek 와 c 같으면 -> pop
                다르면 -> 스택에 넣기
        스택빔 -> 1
        아님 -> 0
        */
        if (s.length() == 1) return 0;
        Deque<Character> stack = new ArrayDeque<>();
        stack.push(s.charAt(0));
        for (char c: s.substring(1).toCharArray()){
            if (stack.isEmpty()){stack.push(c);}
            else {
                if (stack.peek() == c){stack.pop();}
                else {stack.push(c);}
            }
        }
        return stack.isEmpty()? 1: 0;
    }
}

File: 프로그래머스/2/12973. 짝지어 제거하기/짝지어 제거하기.py
def solution(s):
    answer = -1

    """
    모두제거 1
    아니면 0
    
    for:  stack에 하나씩 순서대로 추가함
    b     baa
    """
    if len(s) == 1: return 0
    if len(s) == 2:
        if s[0] == s[1]: return 1
        else: return 0
    
    stack = []
    stack.append(s[0])
    for c in s[1:]:
        if stack and stack[-1] == c:
            stack.pop()
        else:
            stack.append(c)
    if stack: return 0
    else: return 1